react-front-end/tsrc/drm/DrmAcceptanceDialog.tsx

```diff
@@ -0,0 +1,344 @@
+    case "init":
+      return { status: "initialising", dialog: { ...SkeletonDialogProps } };
+    case "termsRetrieved":
+      const {
+        acceptTerms,
+        rejectTerms,
+        drmDetails: { title, subtitle, agreements },
+      } = action;
+      return {
+        status: "successful",
+        dialog: {
+          title,
+          content: (
+            <Grid container>
+              <Grid item>
+                <Typography variant="subtitle1">{subtitle}</Typography>
+              </Grid>
+              <Grid item>
+                <DrmTerms {...agreements} />
+              </Grid>
+            </Grid>
+          ),
+          buttons: (
+            <>
+              <Button onClick={rejectTerms} color="secondary">
+                {languageStrings.common.action.reject}
+              </Button>
+              <Button
+                onClick={(event) => {
+                  acceptTerms();
+                  event.stopPropagation();
+                }}
+                color="primary"
+                autoFocus
+              >
+                {languageStrings.common.action.accept}
+              </Button>
+            </>
+          ),
+        },
+      };
+    case "error":
+      return {
+        status: "failed",
+        dialog: {
+          title: languageStrings.drm.getTerms.error,
+          content: <Typography>{action.cause.message}</Typography>,
+          buttons: (
+            <Button onClick={action.onClose} color="primary">
+              {commonString.action.ok}
+            </Button>
+          ),
+        },
+      };
+    default:
+      throw new TypeError("Unexpected action passed to reducer!");
+  }
+};
+
+/**
+ * Provide a dialog to allow users to accept DRM terms. The dialog has threes states. In initialising stage, it calls the supplied
+ * term provider to retrieve DRM terms and shows a skeleton dialog before terms are retrieved. Then, depending on the call result,
+ * it shows a dialog listing all terms or an error dialog describing why failed.
+ */
+export const DrmAcceptanceDialog = ({
+  termsProvider,
+  onReject,
+  onAccept,
+  onAcceptCallBack,
+  open,
+}: DrmAcceptanceDialogProps) => {
+  const [state, dispatch] = useReducer(reducer, {
+    status: "initialising",
+    dialog: { ...SkeletonDialogProps },
+  });
+
+  // Run the given async task in a TaskEither and return an Either where Left is Error and Right is the specified type T.
+  const runInTask = <T,>(task: () => Promise<T>): Promise<E.Either<Error, T>> =>
+    pipe(TE.tryCatch<Error, T>(task, (cause) => new Error(`${cause}`)))();
+
+  useEffect(() => {
+    const buildActionForError = (cause: Error): Action => ({
+      type: "error",
+      cause,
+      onClose: onReject,
+    });
+
+    // Call 'onAccept' first, and then call 'onAcceptCallBack' or show error message.
+    const acceptDrmTerms = async () => {
+      const result: E.Either<Error, void> = await runInTask(onAccept);
+      pipe(
+        result,
```

Comment:
Are you able to just do:

```suggestion
      pipe(
        await runInTask(onAccept),
```

Or did you break it out for readability? (I note the same done further down.)

Reply:
I was not sure if we can do a `await` in a pipe. Now, I think we can for the first input but maybe can't in the mid of pipe.